src/poetry/core/packages/dependency_group.py

```diff
@@ -26,7 +26,20 @@ def name(self) -> str:
 
     @property
     def dependencies(self) -> list[Dependency]:
-        return self._dependencies or self._poetry_dependencies
+        if not self._dependencies:
```

Comment:
I guess I understood what's going on here :muscle: And it looks like it should work.

As far as I see, we don't take into account the `dynamic` key of the `project` section anywhere in our codebase. Instead we are doing educated guesses. Wouldn't it be easier to rely on this field?

Reply:
At first, I thought it is redundant. However, there is at least one use case where it is necessary:

Projects that do not have any mandatory dependencies but optional dependencies and use `tool.poetry.dependencies` to define sources for these optional dependencies. Without taking `project.dynamic` into account the dependencies in `tool.poetry.dependencies` would have been considered mandatory dependencies in this case.

I added a second commit that takes `project.dynamic` into account to fix this use case.